content/golang/docker/publish-images-multi-arch.md

```diff
@@ -15,45 +15,512 @@ tags:
 }
-{{< / highlight >}}
+{{</highlight>}}
+
+When it is run, it will start a HTTP Server that listens on port `8080`, and will simply respond with the `hello world` text for every request. It is enough for us for now just to make sure everything works fine.
+
+## Testing the Application Without Docker
+
+{{<alert "info">}}
+Since we will use Docker to build the image, this step can be skipped.
+{{</alert>}}
+
+To build and run the application, use
+
+```
+go run .
+```
+
+It will start the server, then you should be able to visit `http://localhost:8080` in your browser, and see the `hello world` text.
+
+
+{{<alert "warn">}}
+If the port is already in use, the server will panic! You can pick a port other than 8080 if you wish, but make sure to use the same port throughout the tutorial.
+{{</alert>}}
+
+# Building with Docker
+
+It is time to build a Docker image with our application in it.
+
+The common practice for Go projects is to separate the build process of the application from the distribution image.
+This approach works better with Continuos Integration and Delivery, allowing more caching opportunities.
+
+However, you could also use [multi-stage](https://docs.docker.com/develop/develop-images/multistage-build/) build with a single Dockerfile if you wanted.
+
+{{<alert "info">}}
+If you are not familiar with Dockerfiles, you can read more about them <a href="https://docs.docker.com/engine/reference/builder/">here</a>.
+{{</alert>}}
+
+## Building for ARM64
+
+First we build the application using Docker with the official Go image.
+
+We will need to set three environment variables for the Go compiler:
+
+- `CGO_ENABLED=0`: This one disables [CGo](https://github.com/golang/go/wiki/cgo), it is only needed in our case because we will distribute with an [Alpine Linux](https://alpinelinux.org/) image, and it would run into problems with linking to `libc`.
+
+
+- `GOOS=linux`: This one sets the target operating system to Linux.
+
+
+- `GOARCH=arm64`: This one sets the target **ARM64** architecture.
+
+{{<alert "info">}}
+You can see the available GOOS and GOARCH combinations <a href="https://golang.org/doc/install/source#environment">here</a>.
+{{</alert>}}
+
+We also set the working directory of the image to `/app`, and mount the current directory there using Docker [volumes](https://docs.docker.com/storage/volumes/).
+
+{{<alert "warn">}}
+If you use cmd on Windows, you must use "-v %cd%:/app" to mount the current directory.
+{{</alert>}}
+
+So our final Docker command will be:
+
+```
+docker run \
+  -e CGO_ENABLED=0 \
+  -e GOOS=linux \
+  -e GOARCH=arm64 \
+  -w /app \
+  -v ${PWD}:/app \
+  golang:1.13 go build
+```
+
+This will produce a `hello-world` application in our project's directory.
+
+Then we will need to distribute it in a Docker image.
+In order to do that, create a `Dockerfile` with the following content:
+
+{{<highlight text "linenos=table,hl_lines=99">}}
+# We use the ARM64 version of Alpine.
+FROM arm64v8/alpine
```

Comment:
The official alpine image provides a manifest, so a simple `FROM alpine` would also work fine if the Docker image gets built on that platform.

Reply:
I added it explicitly, because that part of the tutorial doesn't rely on the host os/arch, but I added notes that it could be omitted in the later parts of the guide.